src/libsync/pushnotifications.cpp

```diff
@@ -0,0 +1,112 @@
+#include "pushnotifications.h"
+#include "creds/abstractcredentials.h"
+#include "websocket.h"
+#include "account.h"
+
+namespace OCC {
+
+PushNotifications::PushNotifications(Account *account, QSharedPointer<AbstractWebSocket> webSocket)
+    : _account(account)
+    , _webSocket(webSocket)
+{
+}
+
+PushNotificationTypes PushNotifications::pushNotificationsAvailable() const
+{
+    return _account->capabilities().pushNotificationsAvailable();
+}
+
+void PushNotifications::reconnect()
+{
+    connectWebSocket();
+}
+
+void PushNotifications::connectWebSocket()
+{
+    disconnectWebSocket();
+
+    setupFilesPushNotifications();
+
+    openWebSocket();
+}
+
+void PushNotifications::disconnectWebSocket()
+{
+    _webSocket->close();
+}
+
+void PushNotifications::onWebSocketConnected()
+{
+    disconnect(_webSocket.data(), &AbstractWebSocket::textMessageReceived, this, &PushNotifications::onWebSocketTextMessageReceived);
+    connect(_webSocket.data(), &AbstractWebSocket::textMessageReceived, this, &PushNotifications::onWebSocketTextMessageReceived);
+
+    authenticateOnWebSocket();
+}
+
+void PushNotifications::authenticateOnWebSocket()
+{
+    const auto credentials = _account->credentials();
+    const auto username = credentials->user();
+    const auto password = credentials->password();
+
+    // Authenticate
+    _webSocket->sendTextMessage(username);
+    _webSocket->sendTextMessage(password);
+}
+
+void PushNotifications::onWebSocketDisconnected() { }
+
+void PushNotifications::onWebSocketTextMessageReceived(const QString &message)
+{
+    if (message == "notify_file") {
+        emit filesChanged(_account);
+    } else if (message == "authenticated") {
+        // TODO: Log
+    } else if (message == "err: Invalid credentials") {
+        connectWebSocket();
```

Comment:
@er-vin @FlexW

Up to you guys, but, I'd add a timeout here (maybe 1 minute?) - if the credentials were invalid, will they become valid immediately?

Reply:
Don't think they would. In such case we should just give up on using push notifications not reconnect. Otherwise there's a high chance we'd get in a disconnect/connect loop and since the timer won't be used during that time the client will never sync the folders of that account on it own.

Reply:
On the first time invalid credentials is sended, push notifications should be disabled?